rust/scx_utils/src/load_balance.rs

```diff
@@ -0,0 +1,318 @@
+    /// infeasibility.
+    pub fn dom_load_sums(&self) -> &[f64] {
+        &self.dom_load_sums
+    }
+
+    /// If applicable, return the adjusted weight for all infeasible scheduling
+    /// entities.
+    pub fn w_prime(&self) -> f64 {
+        self.w_prime
+    }
+}
+
+#[derive(Debug)]
+struct Domain {
+    loads: BTreeMap<usize, f64>,
+    dcycle_sum: f64,
+    load_sum: f64,
+}
+
+fn approx_eq(a: f64, b: f64) -> bool {
+    (a - b).abs() <= 0.0001f64
+}
+
+fn approx_ge(a: f64, b: f64) -> bool {
+    a > b || approx_eq(a, b)
+}
+
+#[derive(Debug)]
+pub struct LoadAggregator {
+    doms: BTreeMap<usize, Domain>,
+    global_loads: BTreeMap<usize, f64>,
+    nr_cpus: usize,
+    max_weight: usize,
+    global_dcycle_sum: f64,
+    global_load_sum: f64,
+    w_prime: f64,
+    dcycle_only: bool,
+}
+
+impl LoadAggregator {
+    /// Given a LoadAggregator with recorded domain loads, compute the
+    /// system-wide load, adjusting for infeasible weights when necessary.
+    pub fn calculate(&mut self) -> LoadBalancer {
+        if !self.dcycle_only && approx_ge(self.max_weight as f64, self.infeasible_threshold()) {
+            self.adjust_infeas_weights();
+        }
+        
+        let mut dom_load_sums = Vec::new();
+        let mut dom_dcycle_sums = Vec::new();
+
+        for (_, dom) in self.doms.iter() {
+            dom_load_sums.push(dom.load_sum);
+            dom_dcycle_sums.push(dom.dcycle_sum);
+        }
+
+        LoadBalancer {
+            dom_load_sums,
+            dom_dcycle_sums,
+            global_dcycle_sum: self.global_dcycle_sum,
+            global_load_sum: self.global_load_sum,
+            w_prime: self.w_prime,
+        }
+    }
+
+    /// Record an instance of some domain's load (by specifying its weight and
+    /// dcycle). Returns an error if duty cycle is specified once for a given
+    /// (Domain, weight) tuple.
+    pub fn record_dom_load(&mut self, dom_id: usize, weight: usize, dcycle: f64) -> Result<()> {
+        if weight < MIN_WEIGHT || weight > MAX_WEIGHT {
+            bail!("weight {} was outside of [{}, {}]", weight, MIN_WEIGHT, MAX_WEIGHT);
+        }
+
+        if !self.doms.contains_key(&dom_id) {
+            self.doms.insert(dom_id, Domain {
+                loads: BTreeMap::new(),
+                dcycle_sum: 0.0f64,
+                load_sum: 0.0f64,
+            });
+        }
+
+        let domain = self.doms.get_mut(&dom_id).unwrap();
+        if let Some(_) = domain.loads.insert(weight, dcycle) {
+            bail!("Domain {} already had load for weight {}", dom_id, weight);
+        }
+
+        if !self.global_loads.contains_key(&weight) {
+            self.global_loads.insert(weight, 0.0f64);
+        }
+        let mut weight_dcycle = self.global_loads.get(&weight).unwrap().clone();
+        weight_dcycle += dcycle;
+        self.global_loads.insert(weight, weight_dcycle);
+
+        if approx_eq(dcycle, 0.0f64) {
```

Comment:
Why do we need this? Would this problematic in configurations which have a lot of weight slots (e.g. compound cgroup * task weights)?

Reply:
When I was implementing infeasible weights in rusty I saw a lot of cases where we'd have a very, very, very low nonzero number that functionally had no effect on load, but still made us do extra work, and even worse, possibly have to deal with issues from finite precision like having a very small negative number when computing `lambda_x`.

Reply:
Regardless of the optimization, the calculations should be safe. I'm not sure it'd be safe to depend on the threshold to protect against that.